elasticsearch_dsl/search_base.py

```diff
@@ -352,20 +352,27 @@ def __getitem__(self, n):
             # If negative slicing, abort.
             if n.start and n.start < 0 or n.stop and n.stop < 0:
                 raise ValueError("Search does not support negative slicing.")
-            # Elasticsearch won't get all results so we default to size: 10 if
-            # stop not given.
-            s._extra["from"] = n.start or 0
-            s._extra["size"] = max(
-                0, n.stop - (n.start or 0) if n.stop is not None else 10
-            )
-            return s
+            slice_start = n.start
+            slice_stop = n.stop
         else:  # This is an index lookup, equivalent to slicing by [n:n+1].
             # If negative index, abort.
             if n < 0:
                 raise ValueError("Search does not support negative indexing.")
-            s._extra["from"] = n
-            s._extra["size"] = 1
-            return s
+            slice_start = n
+            slice_stop = n + 1
+        # Elasticsearch won't get all results so we default to size: 10 if
+        # stop not given.
+        old_from = s._extra.get("from", 0)
+        old_to = old_from + s._extra.get(
+            "size", slice_stop or (slice_start or old_from) + 10
```

Comment:
Why would `old_to` depend on `slice_stop` or `slice_start` which are the new values?

Reply:
I undesrtand that this looks wrong, but I think it is the correct approach for us. The problem is that the initial default of size=10 is a "soft" default that needs to be overriden if the first slicing operation specifies a different size.

Consider this example:

- we start with a default slice of `[0:10]`.
- now you ask for `Search()[:40]`. With a normal slicing operation, you would get `[0:10]` again. What we want is for this first slicing operation to change the size.

I hope this makes sense!

Reply:
> now you ask for `Search()[:40]`. With a normal slicing operation, you would get `[0:10]` again. What we want is for this first slicing operation to change the size.

I don't know, it's not that intuitive. I assumed `s[:10][:40]` was the same as `s[:10]`. Can we add a test about this?

The Django version linked by Honza in the original issue is much simpler: https://github.com/django/django/blob/ca5cd3e3e8e53f15e68ccd727ec8fe719cc48099/django/db/models/sql/query.py#L2116-L2137. Decoupling high/low and from/size and the default size limit sounds smart.

Reply:
> I don't know, it's not that intuitive. I assumed s[:10][:40] was the same as s[:10]. Can we add a test about this?

Yes, this is actually how it works. Just added a couple more tests around this.

The idea here is that `s[:10][:40]` == `s[:10]`, but `s[:40]` should get 40 results, not 10.

The Django code is simpler, I think, because it does not have an implied default of 10 results, which we need to keep for backwards compatibility.